File: NOVEMBER-2025/28-11-2025/2872.ts
function maxKDivisibleComponents(n: number, edges: number[][], values: number[], k: number): number {
    const g: number[][] = Array.from({ length: n }, () => []);

    for (const [a, b] of edges) {
        g[a].push(b);
        g[b].push(a);
    }

    let ans = 0;

    const dfs = (u: number, p: number): number => {
        let sum = values[u];
        for (const v of g[u]) {
            if (v === p) continue;
            sum += dfs(v, u);
        }
        if (sum % k === 0) {
            ans++;
            return 0;
        }
        return sum;
    };

    dfs(0, -1);
    return ans;
}

File: NOVEMBER-2025/28-11-2025/2872.swift
class Solution {
    func maxKDivisibleComponents(_ n: Int, _ edges: [[Int]], _ values: [Int], _ k: Int) -> Int {
        var g = Array(repeating: [Int](), count: n)
        for e in edges {
            let a = e[0], b = e[1]
            g[a].append(b)
            g[b].append(a)
        }

        var ans = 0

        func dfs(_ u: Int, _ p: Int) -> Int64 {
            var sum = Int64(values[u])
            for v in g[u] {
                if v == p { continue }
                sum += dfs(v, u)
            }
            if sum % Int64(k) == 0 {
                ans += 1
                return 0
            }
            return sum
        }

        _ = dfs(0, -1)
        return ans
    }
}